kvserver,snaprecv: move MultiSSTWriter

tbg opened the pull request:
This series of commits moves the multiSSTWriter to newly created package `kvserver/kvstorage/snaprecv`. The plan is that it will be joined by the code extracted in https://github.com/cockroachdb/cockroach/pull/145328, so that we can build a datadriven test that exercises the snapshot receive path's storage logic end to end.

Epic: CRDB-46488
Release note: none

tbg committed:
kvserver: move a method to msstw

tbg committed:
kvserver: move a method

Mechanical.

tbg committed:
snaprecv: incorporate SSTSnapshotStorage and MultiSSTWriter

This mostly mechanically moves SSTSnapshotStorage and MultiSSTWriter to a new
package `snaprecv`. NewMultiSSTWriter now takes an options struct to avoid it
taking to int64s, which can lead to bugs.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/145417)
<!-- Reviewable:end -->

pav-kv commented on pkg/kv/kvserver/multi_sst_writer.go:
```diff
@@ -100,6 +100,78 @@ func newMultiSSTWriter(
 	return msstw, nil
 }
 
+func (msstw *multiSSTWriter) readOneToBatch(
```
Hard to say in advance whether this belongs to `multiSSTWriter`. Logically, this type could be agnostic of where the key/values come from (`batchReader` or something else e.g. inserted manually) and be only concerned with how to write them. Right now the API becomes a mix of low-level `msstw.Put<something>` and high-level "feed a whole batch", and it's easy to misuse.

No strong suggestion on how to do it differently though, it's probably fine as is since this code is coupled anyway and it's quite nice to have it all in one place.

Is there some expectation to capture in a comment, e.g. on the order in which batches are fed into here? Can there be more than one? Does moving this method here make `Put<something>` methods "internal" and they no longer should be intended for use?

tbg replied:
I was able to make all the Put methods internal.

pav-kv commented on pkg/kv/kvserver/multi_sst_writer.go:
```diff
@@ -64,13 +64,18 @@ type multiSSTWriter struct {
 }
 
+type multiSSTWriterOptions struct {
+	SSTChunkSize int64
```
Add comments? Also, seems like these parameters are not covered with tests.
Make it clear what happens when either/all of these params are zero? See another comment below.

tbg replied:
Good point, done.

pav-kv reviewed (commented):
Anything to unexport after the move of `MultiSSTWriter`? E.g. are lower-level put operations intended to remain usable directly?

tbg commented on pkg/kv/kvserver/replica_sst_snapshot_storage_test.go:
That's true, but all tests passed zero explicitly before, and the production code always (i.e. in the one place that makes an MSSTW) reads off the cluster setting. So this should be equivalent.

pav-kv replied:
The test only passes 0 for one param, and the second one is read from a cluster setting (inside `newMultiSSTWriter`). Now, both params are zero.

tbg replied:
Oh - that's what you mean, I see. This is an inconsequential change, since the cluster setting defaults to 100mb, so practically infinite (as far as this test is concerned). There was however a problem in one of the other tests, where the setting was intentionally lowered but I didn't pass this through to the option struct, see the fixup commits.

tbg committed:
snaprecv: make methods private

tbg committed:
snaprecv: add comments on MultiSSTWriterOptions fields

tbg committed:
snaprecv: fix return param name

tbg commented on pkg/kv/kvserver/multi_sst_writer.go:
```diff
@@ -499,21 +499,21 @@ func (msstw *MultiSSTWriter) PutRangeKey(
 }
 
-func (msstw *MultiSSTWriter) Finish(ctx context.Context) (int64, error) {
+func (msstw *MultiSSTWriter) Finish(ctx context.Context) (_dataSize, sstSize int64, _ error) {
```
No, just inconsistent. Fixed.

tbg commented:
I think I addressed all of your comments, RFAL! I also fixed a bunch of stuff; somehow when moving the file with Goland it didn't update the tests, and I didn't notice. There are a lot of fixup commits, once you've reviewed and are happy I will likely end up squashing the initial refactor to avoid spending way too much time making each commit build in isolation.

I'm planning to add a datadriven test for the MSSTW before merging this, but let me know if you'd prefer doing that in a separate PR.

tbg committed:
snaprecv: avoid double counting in msstw

We were previously adding the last open SST's DataSize into EstimatedDataSize()
even if that DataSize was already reflected in `msstw.sstSize`. This led to
double-counting and is now fixed.

tbg committed:
snaprecv: actually use DataSize in EstimatedDataSize

It was previously using a mix of SST size and data size. I got worried that
this might actually regress. Namely, if an SST got finished, the data
previously in DataSize would now be reflected in the size of the SST, but
compression could result in an overall lower number. This is not something the
users of this number would have handled well (we would've passed negative
numbers to the pacer, and as far as I can tell there's no check).

Simply remove this kind of problem by returning the cumulative data size.

As a nice side effect, we also get a taste for both the SST and data size
in the datadriven tests.

tbg commented:
@pav-kv one last round of review please for the last two commits:

- snaprecv: avoid double counting in msstw
- snaprecv: actually use DataSize in EstimatedDataSize

I realized (remembered, really) that we do have basic datadriven test for the MSSTW already, so I'll merge on LGTM and instead spend time on testing `applySnapshotRaftMuLocked`.

tbg commented:
TFTR!

bors r+
